rmd-22906 Add workflow tasks

DavidGamba opened the pull request:
* Instead of using dependencies within tasks, directly call the classes
and pass the required parameters. Allows for easy mocking and testing.

@tonytvo please review.

Commit:
rmd-22906 Add workflow tasks

* Instead of using dependencies within tasks, directly call the classes
and pass the required parameters. Allows for easy mocking and testing.

Commit:
rmd-22906 Remove extra task dependency

Commit:
rmd-22906 Add missing group to Sandbox Workflow task

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -35,21 +35,22 @@ class VeracodeBeginScanTask extends VeracodeTask {
-        Set<String> moduleIds = VeracodePreScanResults.extractWhitelistModuleIds(XMLIO.readXml(preScanResultsOutputFile), getModuleWhitelist())
+        getPreScanResultsTask.app_id = app_id
+        getPreScanResultsTask.veracodeAPI = veracodeAPI
+        getPreScanResultsTask.run()
```
- there are couple concerns I have with this:
  - this task reach too deep and depends on another gradle task (would it be easier to test if the prescanresults behavior is isolated in domain model instead of gradle tasks?)
  - and this gradle task would call the behavior of that PreScanResults to get the pre-scan results behavior?
  - run(), what does run do? run is the API of the class, it is the method that would be used by other classes, should it be named to something that has explicit/descriptive meaning?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodePlugin.groovy:
```diff
@@ -34,6 +34,10 @@ class VeracodePlugin implements Plugin<Project> {
         project.extensions.create('veracodeSetup', VeracodeSetup)
 
+        // Main tasks
+        project.task(VeracodeWorkflowTask.NAME, type: VeracodeWorkflowTask)
```
what does the vercodeworkflowtask accomplish? we might want a more descriptive name here....

DavidGamba replied:
It does the following:
* create build
* upload files
* begin pre-scan
* begin scan

And it gets the build status in between each step.

I couldn't come up with a better name so I would appreciate suggestions.

tonytvo replied:
it seems to me we can abstract it to something like submitToVeracodeScan? veracodeAnalysis?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+        build_version = project.findProperty("build_version")
+    }
+
+    VeracodeGetBuildInfoSandboxTask buildInfoSandboxTask = new VeracodeGetBuildInfoSandboxTask()
```
this gradle tasks depends on a whole bunch of other gradle tasks is a little bit worrisome to me....

DavidGamba replied:
In what way is dependency worrisome? Can you be more explicit please?

This task is a task that runs the entire workflow so it basically orchestrates running the other tasks. I would not want to duplicate that behaviour. But I don't really know what your concern is.

tonytvo replied:
- sorry, I didn't clear with my concerns, as far as I understand, gradle custom task were not designed to compose other tasks, in case you need to do it, you can use aggregate tasks with dependsOn....
- what I'm getting at is you're probably aware that I'm not a big fan of putting too much logic/stuff into gradle custom task class as this is the API between gradle and our custom business logic. It would make it harder to test custom business logic
  - if we include business logic in the gradle custom task, then it would rely on gradle support such as testkit or gradle tools to support testing gradle custom tasks. 
  - What I have seen over the years is recommendation to separate the domain logic from technology details as much as possible. The main reason for that is business logic has to live longer than the technology, what if gradle deprecated, what if we want to use something else other than gradle.
  - We should build our product in such a way that it's easy to switch to different technologies.....

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+
+        // Get build info
+        printf "[debug] getBuildInfo\n"
+        buildInfoSandboxTask.run()
```
I'm not a big fan of field/property injection as it's not explicit what's require to use the object.
- couple alternatives I could see for buildInfoSandboxTask.run()
  - new GetBuildInfoSandbox(sandbox_id, app_id).getBuildInfo()
  - buildInfoSandbox.getBuildInfo(sandbox_id, app_id)

DavidGamba replied:
I can't call the constructor within a method of a gradle task. It has to be as a global so gradle can figure our dependencies and such. That is why I couldn't do something like that.

Also, gradle plugins depend on external properties passed through the command line or defined on the gradle.properties. Injecting the properties seems to me like it aligns with what I have been doing so far across the entire plugin.

tonytvo replied:
"can't call constructor" within a method is a smell for me -> it sounds like we depends on gradle on how it figures out dependencies and properties so much that we can't call constructor in the method. Again, if you extract domain logic away from gradle custom task, you would have more flexibility here.....

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+            createBuildSandboxTask.run()
+            buildStatus = buildInfoSandboxTask.getBuildStatus()
+            printf "[debug] buildStatus: %s\n", buildStatus
+        }
```
should we stop here, or the buildStatus could be incomplete and the prescan would happen?

DavidGamba replied:
Yes, the `createBuild` task is immediate so the buildstatus right after changes to `incomplete`.

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+        String buildStatus = buildInfoSandboxTask.getBuildStatus()
+        printf "[debug] buildStatus: %s\n", buildStatus
+
+        // Done previous work
```
hmmm.....  based on the comment, it sounds like you unconsciously want to extract the if block to the function and name it as createBuildSandboxIfPreviousWorkIsDone

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+            printf "[debug] buildStatus: %s\n", buildStatus
+        }
+
+        // Clean build or with some uploaded files
```
Again based on the comments, you are looking at extract the if block to something like
uploadFileAndBeginPreScanForCleanBuild?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:
```diff
@@ -0,0 +1,125 @@
+            printf "[debug] buildStatus: %s\n", buildStatus
+        }
+
+        // Pre-Scan completed
```
again, sounds like you want to extract if block to the function named: beginScanIfPreScanSucceeded?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowTask.groovy:
```diff
@@ -0,0 +1,115 @@
+    VeracodeBeginPreScanTask beginPreScanTask = new VeracodeBeginPreScanTask()
+    VeracodeBeginScanTask beginScanTask = new VeracodeBeginScanTask()
+
+    void run() {
```
hmmmm.... there's quite a bit of duplication here? some duplication removal might comes in handy here !!!

DavidGamba committed:
rmd-22906 Update README with workflow task information

DavidGamba committed:
rmd-22906 Move Output file definition from task to Domain Object

DavidGamba committed:
rmd-22906 Refactor uploadFile tasks

Move logic to the Domain Object

DavidGamba committed:
rmd-22906 Refactor BeginScan tasks

Get PreScan Results from the Domain Object instead of instantiating a
gradle task.

DavidGamba committed:
rmd-22906 Refactor workflow tasks

* Add a Domain Object for task logic.
* Don't instantiate other tasks.
* Add ignoreFailure property.

DavidGamba committed:
rmd-22906 Remove getBuildStatus method from task

DavidGamba commented:
@tonytvo please review. Please let me know if the size is OK or if you prefer to have it split in smaller PRs.
